---
title: "Single cell RNAseq unimodal Seurat based workflow for droplet single cell sequencing"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("C:/Users/aflorescu/Molecular Partners AG/DEV_TP_ExVivo - Ana/Rscripts/SingleCellSeuratBased/scRNAseq-unimodal-Seurat-based-functions.R")
library(tidyverse)
library(Seurat)
library(scran)
library(DropletUtils)
```

## Introduction

This is a workflow for a standardized analysis for single cell RNAseq data (no other modalities). Analysis is based on Seurat object, and several wrapper function have been made. The functions are stored in a separate file. As this is pubclid data, the workflows starts form a count file or a 10x output (barcode, feature and matrix file). 


## Read data, do QCs, remove empty droplets (correction of ambient RNA and inference of doublets from RNA only are not yet included)

```{r}
#read data 
#remove empty drops
counts.clean <- RemoveEmptyDrops.Barcodes(count,matrix)
#Create Seurat object
# Remove cells with high or low gene counts and/or high mitochondrial genes
SeuratObj[["percent.mt"]] <- PercentageFeatureSet(SeuratObj, pattern = "^MT-")
#SeuratObj <- AddMetaData(SeuratObj, MetaData)
SeuratObj <- subset(SeuratObj, subset = nFeature_RNA<4000 & nFeature_RNA > 300 & percent.mt < 10) 
```


## Nromalization, dimesnionality reduction and clustering 

The prefered  method is SCTransform.Compared to the "usual" Seurat workflow the SC Transfrom workflow does not require find variable features, as they are directly returned by SCTransform and so is scale data. 

```{r}
SeuratObj <- SCTransform(SeuratObj,  vars.to.regress = "percent.mt", verbose = FALSE)
SeuratObj <- RunPCA(object = SeuratObj)
SeuratObj <- FindNeighbors(SeuratObj, dims = 1:30) 
SeuratObj <- FindClusters(object = SeuratObj, resolution=clusterRes)
#SeuratObj1 <- AddMetaData(SeuratObj1, MDsubset)
#SeuratObj <- RunUMAP(SeuratObj, dim = 1:10)
DimPlot(SeuratObj, reduction = "umap")
```



## Data integration (bacth effect)

If data comes from multiple batches visual inspection (e.g. UMAP colored by batches and cell types) is recommendedn. If it looks like there are batch effected , dataset alignment is necessary. 
This workflow uses Harmony, based on two benchmark papers (Tran et al, Genome Biology 2020 and a preprint from the Theis lab: Luecher et al, Bioarxiv). The Harmony reference paper is Kolsunsky et al, Nature Methods, 2019. 



## Cell cylce markers and scores

We can also compute a gene module score for genes that are related to the cell cycle and assign each cell to a phase, which will appear as part of the annotation. Seurat includes the required gene signatures in "cc.genes". 

```{r}
SeuratObj <- CellCycleScoring(SeuratObj, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes, set.ident = TRUE)
```


##Cell typing

Cell typing will be done in two rounds. First the main clusters are labelled using SingleR and then the immune cells  are  re-typed using Azimuth.Azimuth is a classifier based on a reference CITE-seq single cell dataset developed by Satija lab. However, since Azimuth is PBMC based it will not recognize  all non-immune cell types (tumor, endothelial cells, fibroblasts) and granulocytes.  

If data is from PBMCs using Azimuth is enough.

If data is from PDAC, Azimuth now has a pancreas reference (see  on line app and Seurat vignette). However, the workflow seems to be a bit different as data is from multiple technologies. 
 


## For cancer data: call tumor cells using inferCNV


## Trajectory inference (for a group of cells)

## Cell cell communication - 


## Other newtork or environment inference ? 

## Appendix - reference- based cell typing for pancreas
