---
title: "Single cell RNAseq unimodal Seurat based workflow for droplet single cell sequencing"
output: html_document
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
source("C:/Users/aflorescu/Molecular Partners AG/DEV_TP_ExVivo - Ana/Rscripts/SingleCellSeuratBased/scRNAseq-unimodal-Seurat-based-functions.R")
library(tidyverse)
library(Seurat)
library(scran)
library(DropletUtils)
library(cowplot)
library(data.table)
library(biomaRt)
```

## Introduction

This is a workflow for a standardized analysis for single cell RNAseq data (no other modalities). Analysis is based on Seurat object, and several wrapper function have been made. The functions are stored in a separate file. As this is public data, the workflows starts form a count file or a 10x output (barcode, feature and matrix file). 


## Read data, do QCs, remove empty droplets (correction of ambient RNA and inference of doublets from RNA only are not yet included)

The removal of empty drops done here via the DropletUtils package is an alternate way of doing what the CellRanger filter function does. This mean that it does not have to be ran on filtered 10x matrix (it actually gives errors because it requires low count columns to compute ambient RNA profile and those are removed by 10x filtering ). This should be ran on public data that is not cleaned prior. 

Define first a folder path
```{r}
fpath <- "H:/data/10x datasets/GSE137829_Prostate_neuroendocrine"
datasetname  <- "P5"
```


```{r}
#read data - example
#count.matrix <- Read10X(data.dir = paste(fpath, datasetname,sep="/"))
#count.matrix <- readRDS("H:/data/10x datasets/To-process/brain-reference.rds")
count.matrix <- fread("H:/data/10x datasets/GSE137829_Prostate_neuroendocrine/GSM4711414_P5_gene_cell_exprs_table.txt", sep="\t", data.table=FALSE, header=TRUE)
rownames(count.matrix) <- make.unique(count.matrix[,2])
count.matrix <- count.matrix[,-c(1,2)]

```


```{r}
#remove empty drops (this should be run on the unfiltered 10x output, otherwise might say there are no counts to estimate ambient profile)
#counts.clean <- RemoveEmptyDrops.Barcodes(count.matrix)
#Create Seurat object

SeuratObj <- CreateSeuratObject(counts=count.matrix)
rm(count.matrix)
gc()
# Remove cells with high or low gene counts and/or high mitochondrial genes
SeuratObj[["percent.mt"]] <- PercentageFeatureSet(SeuratObj, pattern = "^MT")
VlnPlot(SeuratObj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
#SeuratObj <- AddMetaData(SeuratObj, MetaData)
SeuratObj <- subset(SeuratObj, subset = nFeature_RNA < 10000 & nFeature_RNA > 1000 & percent.mt < 10) 
#saveRDS(SeuratObj, "C:/data/10x datasets/To-proces/PDAC-Broad-radiation/untreated/PDAC-broad-radiation-treated-for-joining.rds")
```


## Normalization, dimensionality reduction and clustering 

The preferred  method is SCTransform.Compared to the "usual" Seurat workflow the SC Transfrom workflow does not require to find variable features, as they are directly returned by SCTransform and so is scale data. 

```{r}
clusterRes <- 0.5 #define and maybe vary a bit cluster resolution depend on the data set. 0.5 is a usually good first option
#SeuratObj <- NormalizeData(SeuratObj, normalization.method = "LogNormalize", scale.factor = 10000)
SeuratObj <- FindVariableFeatures(SeuratObj)
#SeuratObj <- ScaleData(SeuratObj)
SeuratObj <- SCTransform(SeuratObj,  vars.to.regress = "percent.mt", verbose = FALSE)
SeuratObj <- RunPCA(object = SeuratObj)
SeuratObj <- FindNeighbors(SeuratObj, dims = 1:30) 
SeuratObj <- FindClusters(object = SeuratObj, resolution=clusterRes)
SeuratObj <- RunUMAP(SeuratObj, dim = 1:30)
DimPlot(SeuratObj, reduction = "umap")

```



## Data integration (batch effect removal)

If data comes from multiple batches visual inspection (e.g. UMAP colored by batches and cell types) is recommended. If it looks like there are batch effected , dataset alignment is necessary. 
This workflow uses Harmony, based on two benchmark papers (Tran et al, Genome Biology 2020 and a preprint from the Theis lab: Luecher et al, Bioarxiv). The Harmony reference paper is Kolsunsky et al, Nature Methods, 2019. 
This requires that the two dataets were first added to a single Seurat object on which the workflow up to PCA has been run



## Cell cycle markers and scores

We can also compute a gene module score for genes that are related to the cell cycle and assign each cell to a phase, which will appear as part of the annotation. Seurat includes the required gene signatures in "cc.genes". 

```{r}
SeuratObj <- CellCycleScoring(SeuratObj, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes, set.ident = TRUE)
```


##Cell typing

Cell typing will be done in two rounds. First the main clusters are labelled using SingleR and then the immune cells  are  re-typed using Azimuth.Azimuth is a classifier based on a reference CITE-seq single cell dataset developed by Satija lab. However, since Azimuth is PBMC based it will not recognize  all non-immune cell types (tumor, endothelial cells, fibroblasts) and granulocytes.  Azimuth will call everything as PBMC. 

If data is from PBMCs using Azimuth is enough.

If data is from PDAC, Azimuth now has a pancreas reference (see  on line app and Seurat vignette). However, the workflow seems to be a bit different as data is from multiple technologies and is not yet implemented. 
 
```{r}
#SeuratObj <- readRDS("C:/data/10x datasets/Seurat objects/PDAC-Broad-radiation-untreated.rds")
df.encode <- readRDS("C:/Users/aflorescu/Molecular Partners AG/DEV_TP_ExVivo - Ana/ToTransfer/encode-personalized-celltypes.rds")
#run singleR
SeuratObj <- Seurat.Singler.SCT(SeuratObj, df.encode)
#run Azimuth
SeuratObj <- Seurat.Azimuth.celltypes(SeuratObj)

#mix Azimuth and Seurat annotations
MetaData <- SeuratObj@meta.data
MetaData$cell.type <- MetaData$cell.type
MetaData$cell.type.fine <- MetaData$cell.type.detailed
for (i in 1:nrow(MetaData)){
#if ((MetaData$cell.type[i]=="unclassified")) MetaData$cell.type[i]=MetaData$cluster.type.singler[i]
if ((MetaData$cluster.type.singler[i] %in% c("endothelial cell", "epithelial cell", "fibroblast", "neutrophil", "muscle cell"))){ MetaData$cell.type[i]=MetaData$cluster.type.singler[i]
MetaData$cell.type.fine[i]=MetaData$cluster.type.singler[i]

}

if (is.na(MetaData$cell.type.fine[i])) MetaData$cell.type.fine[i]=MetaData$cluster.type.singler[i]
if (MetaData$cell.type.fine[i] == "plasmacytoid dendritic cell") MetaData$cell.type[i] = "plasmacytoid dendritic cell"
if (MetaData$cell.type.fine[i] == "plasmablast") MetaData$cell.type[i] = "plasmablast"
if (MetaData$cell.type.fine[i] == "erythrocyte") MetaData$cell.type[i] = "erythrocyte"
if (MetaData$cell.type.fine[i] == "regulatory T cell") MetaData$cell.type[i] = "regulatory T cell"
if (MetaData$cell.type.fine[i] == "platelet") MetaData$cell.type[i] = "platelet"

}

MetaData$cell.type <- recode(MetaData$cell.type, "monocyte"="monocyte/macrophage", "Mono"="monocyte/macrophage")
SeuratObj <- AddMetaData(SeuratObj, MetaData)

#remove some of the annotations to clean up the files 
SeuratObj@meta.data$predicted.celltype.l1 <- NULL
SeuratObj@meta.data$predicted.celltype.l2 <- NULL
SeuratObj@meta.data$predicted.celltype.l1.score <- NULL
SeuratObj@meta.data$predicted.celltype.l2.score <- NULL
SeuratObj@meta.data$cell.type.detailed <- NULL

#remove prediction assays from Azimuth
SeuratObj[["prediction.score.celltype.l1"]]<- NULL
SeuratObj[["prediction.score.celltype.l2"]]<- NULL
SeuratObj[["predicted_ADT"]]<- NULL

SeuratObj <- SetIdent(SeuratObj, value="cell.type")
p2 <- DimPlot(SeuratObj, reduction = "umap")
SeuratObj <- SetIdent(SeuratObj, value="cell.type")
#p1 <- DimPlot(SeuratObj, reduction = "umap")
p1 <- FeaturePlot(SeuratObj, features = "DLL3")
plot_grid(p1,p2)


#saveRDS(SeuratObj,"H:/data/10x datasets/Seurat objects/brain-reference-log-norm.rds")
```

Read cell metadata from file if available 
```{r}

#MetaData <- readRDS("H:/data/10x datasets/To-process/brain-reference-metadata.rds")
#MetaData <- read.table("H:/data/10x datasets/To-process/brain-reference-metadata.csv", header=TRUE, sep=",", comment.char = "")
#rownames(MetaData) <- MetaData[,2]
#MetaData <- MetaData@meta.data
#SeuratObj <- AddMetaData(SeuratObj, MetaData[,c(3,6,9)])

#MetaData1 <- read.table("H:/data/10x datasets/To-process/GSE182434_DLBCL/GSE182434_cell_annotation.txt", sep="\t", header=TRUE) 


#colnames(MetaData1)[1]<-"CellID"
#colnames(MetaData1)[4]<-"cell.type.original"
#colnames(MetaData1)[2]<-"PatientID"
#colnames(MetaData1)[6]<-"Molecular.subtype"
#colnames(MetaData1)[7]<-"CANCER_TYPE_SHORT"
#MetaData1$CANCER_TYPE_SHORT<- recode(MetaData1$CANCER_TYPE_SHORT, "tonsil"="normal tonsil")

#rownames(MetaData1) <- MetaData1$CellID
MetaData <- SeuratObj@meta.data
MetaData$PatiemtID <- "Patient 6"
MetaData$CANCER_TYPE <- "prostate neuroendocrine carcinoma"
MetaData$CANCER_TYPE_DETAILED <- "prostate neuroendocrine carcinoma"
MetaData$CANCER_TYPE_SHORT <- "PRNE"
MetaData$SAMPLE_TISSUE_OF_ORIGIN <- "prostate gland"
MetaData$SAMPLE_TYPE <-"relapsed tumor transfromed to CRPC"
MetaData$Dataset <- "GSE137829"
MetaData$Age <- 70
MetaData$`PSA_level_at_initial_diagnostic[ng/ml]` <- 72.72
MetaData$Gleason_score_at_initial_diagnostic <- "small cell NEPC"
MetaData$Stage_at_diagnosis <- "cT4N1M1"
MetaData$First_line_therapy <- "Docetaxel"
MetaData$Second_line_therapy <- "None"
MetaData$`Time_from_treatment_start_to_CRPC[months]` <- 14.8
MetaData$`Time_from_CRPC_to_biopsy[months]` <- 3.7
MetaData$`PSA_level_at_biopsy[ng/ml]` <- 7.18
MetaData$Stage_at_biopsy <- "	cT3bN0M1"

#MetaData <- MetaData[,c(2,3,4,7,8,9,15:22,23,24:26,27)]
SeuratObj<- AddMetaData(SeuratObj, MetaData)
#SeuratObj <- subset(SeuratObj, timepoint=="post")
#Seurat requires cell barcodes as rownames
```


Add BCR/TCR seq to meta-data if available:
```{r}

#MetaData1 <- SeuratObj@meta.data

MetaData.bcr <- read.table("H:/data/10x datasets/To-process/GSE182434_DLBCL/DLBCL002B_clonotypes.txt", sep="\t", header=TRUE) 
colnames(MetaData.bcr)[1]<-"CellID"
MetaData.bcr$CellID <- gsub("-1","",MetaData.bcr$CellID)
MetaData.bcr$CellID <- paste(MetaData.bcr$CellID, "DLBCL002B", sep="_")
MetaData.bcr$raw_clonotype_id <- paste(MetaData.bcr$raw_clonotype_id, "DLBCL002B", sep="_")
MetaData2 <- MetaData.bcr

MetaData.bcr <- read.table("H:/data/10x datasets/To-process/GSE182434_DLBCL/DLBCL007B_clonotypes.txt", sep="\t", header=TRUE) 
colnames(MetaData.bcr)[1]<-"CellID"
MetaData.bcr$CellID <- gsub("-1","",MetaData.bcr$CellID)
MetaData.bcr$CellID <- paste(MetaData.bcr$CellID, "DLBCL007B", sep="_")
MetaData.bcr$raw_clonotype_id <- paste(MetaData.bcr$raw_clonotype_id, "DLBCL007B", sep="_")
MetaData2 <- bind_rows(MetaData2, MetaData.bcr)

MetaData.bcr <- read.table("H:/data/10x datasets/To-process/GSE182434_DLBCL/DLBCL008B_clonotypes.txt", sep="\t", header=TRUE) 
colnames(MetaData.bcr)[1]<-"CellID"
MetaData.bcr$CellID <- gsub("-1","",MetaData.bcr$CellID)
MetaData.bcr$CellID <- paste(MetaData.bcr$CellID, "DLBCL008B", sep="_")
MetaData.bcr$raw_clonotype_id <- paste(MetaData.bcr$raw_clonotype_id, "DLBCL008B", sep="_")
MetaData2 <- bind_rows(MetaData2, MetaData.bcr)

MetaData.bcr <- read.table("H:/data/10x datasets/To-process/GSE182434_DLBCL/DLBCL111B_clonotypes.txt", sep="\t", header=TRUE) 
colnames(MetaData.bcr)[1]<-"CellID"
MetaData.bcr$CellID <- gsub("-1","",MetaData.bcr$CellID)
MetaData.bcr$CellID <- paste(MetaData.bcr$CellID, "DLBCL111B", sep="_")
MetaData.bcr$raw_clonotype_id <- paste(MetaData.bcr$raw_clonotype_id, "DLBCL111B", sep="_")
MetaData2 <- bind_rows(MetaData2, MetaData.bcr)


MetaData1 <- merge(MetaData1, MetaData2, by="CellID", all=TRUE)
rownames(MetaData1) <- MetaData1$CellID
MetaData1$CellID.1<-NULL
SeuratObj<- AddMetaData(SeuratObj, MetaData1)
```

Write outputs to file: 

```{r}
write10xCounts(x = SeuratObj@assays$RNA@counts, path = paste(fpath, datasetname, sep="/"), version="3")
filemeta <- paste(datasetname, "cell-metadata.txt", sep="-")
write.table(SeuratObj@meta.data,paste(fpath, datasetname,filemeta, sep="/"), sep="\t", row.names = TRUE)
```

## For cancer data: call tumor cells using inferCNV

InferCNV is a mtehod used to distinguish tumor cells from nomral cells. This method sorts genes by chromosome and genomic position and compares the average gene expression over genomic region to a reference. It uses a window size of 100, but larger window sizes can make sense, depending on the number of genes in your dataset. Ideally we would need normal cells as reference: sometimes people use fibroblast or endothelial cell or even immune cells if a normal control is  not available. 

Rather than creating a wrapper function it seems simpler to source a separate script. This is due to the fact that the workflow is quite complex and is especially time and memory consuming, so it might need to be run separately on another machine or with data split by patients.


We prepare the input and write to files, and then inferCNV will be ran on SevenBridges

```{r}
ref_group_names=c("B cell","CD4 T cell", "CD8 T cell", "endothelial cell", "fibroblast",  "monocyte/macrophage", "natural killer cell")
tumor_group_names=c("epithelial cell", "other")

SeuratObj <- subset(SeuratObj, cell.type %in% c(ref_group_names, tumor_group_names))

raw_counts_matrix <- as.matrix(SeuratObj@assays[["RNA"]]@counts)

# 2. cell annotation files

MetaData <- SeuratObj@meta.data

#------------------------------------------------------
# 3. gene order file (order on Chromosomes)
#------------------------------------------------------
all.genes <- data.frame(hgnc_symbol=rownames(raw_counts_matrix))

# retrieve chromosomes positions given a list of genes 

ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
results_id <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id", 'chromosome_name', 'start_position','end_position'),filters = "hgnc_symbol", values = all.genes$hgnc_symbol, mart = ensembl)

chromo_list <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
                 "11", "12", "13", "14", "15", "16", "17", "18", "19",
                 "20", "21", "22", "X", "Y")

results <- results_id %>% filter(chromosome_name %in% chromo_list)
results <- results %>% dplyr::select(hgnc_symbol, chromosome_name, start_position, end_position)

#check if any duplicates in gene position
rep_gene <- data.frame(table(results$hgnc_symbol))
results[results$hgnc_symbol %in% rep_gene[rep_gene$Freq>1, ]$Var1, ]

#clear replicates
results_unique <- results[!duplicated(results$hgnc_symbol), ]

# write table of gene notations
write.table(results_unique, paste(fpath, datasetname,"-gene_annotation.txt",sep="/"), row.names = FALSE,  sep = "\t", quote = FALSE, col.names=FALSE)

# filter the counts matrix according to results of chromosome positions
counts_matrix <- raw_counts_matrix[c(results_unique$hgnc_symbol), ]
write.table(counts_matrix,  paste(fpath, datasetname,"-count_matrix.txt",sep="/"),  sep = "\t", col.names= TRUE, row.names = TRUE)
#df <- fread("H:/data/inferCNV/GSE131907-batch1-count_matrix.txt", header=TRUE, sep="\t",data.table = FALSE)

#MetaData$Sample_type_patient <- paste(MetaData$Sample_type, MetaData$orig.ident, sep="-")
MetaData$Cell_ID <- rownames(MetaData)
write.table(MetaData[,c(which(colnames(MetaData) %in% c("Cell_ID")), which(colnames(MetaData) %in% c("cell.type")))],paste(fpath, datasetname, "-metadata-for-infer-cnv.txt", sep="/"), row.names = FALSE, sep="\t", quote=FALSE, col.names = FALSE)

```




## Trajectory inference (for a group of cells)

## Cell cell communication - 


## Other newtork or environment inference ? 

## Appendix - reference- based cell typing for pancreas
